src/taco-utils/commands.ts

```diff
@@ -77,36 +82,50 @@ module TacoUtility {
+            }
+
             /**
              * Concrete implementation of ICommand's run
              * Parse the arguments using overridden parseArgs, and then select the most appropriate subcommand to run
              */
-            public run(data: ICommandData): Q.Promise<any> {
-                var deferred: Q.Deferred<any> = Q.defer();
-                this.data = this.parseArgs(data.original);
-
+            protected runCommand(data: ICommandData): Q.Promise<ICommandTelemetryProperties> {
                 // Determine which subcommand we are executing
-                this.executedSubcommand = this.getSubCommand(this.data);
-                if (this.executedSubcommand) {
-                    return this.executedSubcommand.run(this.data).then((telemetryProperties: telemetryHelper.ICommandTelemetryProperties) => {
-                        telemetryProperties["subCommand"] = telemetryHelper.TelemetryHelper.telemetryProperty(this.executedSubcommand.name, /*isPii*/ false);
-                        return telemetryProperties;
-                    });
-                } else {
-                    return Q.reject(errorHelper.get(TacoErrorCodes.CommandBadSubcommand, this.name, this.data.original.toString()));
+                var subCommand = this.getSubCommand(this.data);
+                if (subCommand) {
+                    return subCommand.run(this.data)
+                        .then((telemetryProperties: telemetryHelper.ICommandTelemetryProperties) => {
+                            telemetryProperties["subCommand"] = telemetryHelper.TelemetryHelper.telemetryProperty(subCommand.name, /*isPii*/ false);
+                            return telemetryProperties;
+                        });
                 }
-
-                return deferred.promise;
+                return Q.reject<ICommandTelemetryProperties>(errorHelper.get(TacoErrorCodes.CommandBadSubcommand, this.name, this.data.original.toString()));
             }
 
             public resolveAlias(subCommand: string): string {
                 return (this.info.aliases && this.info.aliases[subCommand]) ? this.info.aliases[subCommand] : subCommand;
             }
 
-            private getSubCommand(options: ICommandData): ICommand {
+            private getSubCommand(options: ICommandData): ISubCommand {
+                // first do a simple name match
+                var name: string = options.remain[0];
+                name = name ? this.resolveAlias(name) : name;
                 for (var i: number = 0; i < this.subcommands.length; ++i) {
-                    var subCommand: ICommand = this.subcommands[i];
-                    if (subCommand.canHandleArgs(options)) {
+                    if (name === this.subcommands[i].name) {
+                        return this.subcommands[i];
+                    }
+                }
+                // if subcommand has specified a canHandleArgs, check if there is a subcommand
+                // which can handle the args
```

Comment:
I'm not sure I understand the intent behind this logic... Why would we try to run a subcommand if the name doesn't match? Does this mean that if `taco kit select` implements `canHandleArgs()` and we do `taco kit dasdasdsa --cordova 5.2.0`, we will end up running `select`? If that's the case, then I feel it might be a bit dangerous to be so permissive